### PR TITLE
Add sensitive custom fields

### LINKED ISSUE #3342: Access control for certain custom fields
Certain custom fields should only be visible to people in an authorizationGroup and counterparts (in case of People) and to people in an authorizationGroup and authors (in the case of Reports).

We could add an optional `authorizationGroups` parameter under `customField`, that specifies the list of uuids of authorization groups.

We will also need some visual indicator on these fields to convey that they are not visible to everyone.

Maybe, it is wiser to store the uuid of the autorization group in the json of the value itself to ensure future access control.

Steps to resolve this issue:

- [ ] Define User Stories.
- [ ] Write all requirements and clarify unclear requirements.
- [x] Write up the architecture of the data and the security rules that apply to it according to the requirements.
- [x] Create the data model (server-side)
- [x] Implement the security rules (server-side)
- [x] Add the sensitive custom fields to the objects that need them (server-side)
- [x] Add some sensitive custom fields to the demo data (server-side) and add tests (server-side)
- [ ] Mock-up of how the sensitive custom fields would work on the client-side.
- [ ] Retrieve sensitive custom fields with an object from the server (report is the most likely target) and show them on the client-side.
- [ ] Edit sensitive custom fields with an object on the client-side and update them on the server.
- [ ] Add client-side tests.

### PR DESCRIPTION
**Pull request by anilsonmez-simsoft**
Person objects have `customSensitiveInformation` which are only visible to admins, counterparts, and users whose position is in the `authorizationGroup`. Admins can edit all sensitive fields while superusers can edit only the ones they are authorized for. 

Closes #3342 

#### User changes
- Authorized users can see sensitive information of people.

#### Super User changes
- Superusers can see and edit sensitive information of people if they are authorized.

#### Admin changes
- Admins can see and edit any sensitive information of people.

#### System admin changes
- [x] anet.yml or anet-dictionary.yml needs change
  Sensitive fields should be added to the `person` and  `showPageOrderedFields` in the dictionary:
  ```
  fields:
    person:
      customSensitiveInformation:
        birthday:
          type: date
          label: Date of birth
          authorizationGroupUuids: [...]
          tooltipText: "This field contains sensitive information!"

    …

    principal:
      person:
        showPageOrderedFields:
          …
          - birthday
          …
  ```
- [x] db needs migration
- [ ] documentation has changed
- [x] graphql schema has changed

### Checklist
  - [x] Described the user behavior in PR body
  - [x] Referenced/updated all related issues
  - [x] commits follow a `repo#issue: Title` title format and [these 7 rules](https://chris.beams.io/posts/git-commit/)
  - [x] commits have a [clean history](https://epage.github.io/dev/commits/), otherwise PR may be squash-merged
  - [x] Added and/or updated unit tests
  - [ ] Added and/or updated e2e tests
  - [x] Added and/or updated data migrations
  - [ ] Updated documentation
  - [x] Resolved all build errors and warnings
  - [x] Opened debt issues for anything not resolved here

### CODE REVIEW

**Review by MehmetAli-Simsoft — APPROVED** [review]

File: client/src/components/Model.js [hunk 1]
```diff
@@ -707,4 +721,31 @@ export default class Model {
...
+  }
+
+  static getAuthorizedSensitiveFields(
+    isAuthorizedCallback,
+    user,
+    customSensitiveInformation,
+    ...args
+  ) {
+    const authorizedFieldsConfig = {}
+    Object.entries(customSensitiveInformation).forEach(([k, v]) => {
```
Comment (asafkaganbezgin) [inline client/src/components/Model.js:745]: More descriptive names could be used for k & v.

**Comment by asafkaganbezgin** [comment]
image

Black font color on darkish blue is makes the text harder to read. Not a vital issue though.

File: client/src/components/Model.js (hunk 1 above)
Comment (anilsonmez-simsoft) [reply client/src/components/Model.js:745]: IMHO it is clear enough that it just creates a new object from the input by filtering some entries. So the return value is just a part of `customSensitiveInformation`.

**Comment by anilsonmez-simsoft** [comment]
> image
> 
> Black font color on darkish blue is makes the text harder to read. Not a vital issue though.



> image
> 
> Black font color on darkish blue makes the text harder to read. Not a vital issue though.

This dark blue color can be configured in the dictionary but maybe we can change it in our dev environment too.

**Review by asafkaganbezgin — APPROVED** [review]

**Comment by gjvoosten** [comment]
> This dark blue color can be configured in the dictionary but maybe we can change it in our dev environment too.

Changed the colours in the demo dictionaries to ones that have a black font with the most contrast ( with guidance from https://www.w3schools.com/cssref/css_colors.asp ).

**Pull request merged**